boa_engine/src/string.rs

```diff
@@ -358,57 +767,71 @@ impl JsString {
         let x = x.as_ref();
         let y = y.as_ref();
 
-        let this = Self {
-            inner: Inner::concat_array(&[x, y]),
-            _marker: PhantomData,
-        };
+        let inner = Inner::concat_array(&[x, y]);
```

Comment:
There is a memory leak here. If the concatenated string is in the `CONSTANTS` (like `"Obj" + "ect"`) it will leak the heap allocated string (since no destructor is called) and return the static string.

This applies to `JsString::concat_array()`